300 converting unit test gdoc to markdown

Dylan-Celli opened the pull request.

samarth9008 commented:
FYI, You should add @DanilYachmenev as the reviewer for now.

gpsaggese reviewed (commented):
I've put together a flow that automates most of the conversion

https://github.com/sorrentum/sorrentum/blob/master/docs/Documentation_about_guidelines.md#gdocs---markdown-using-pandoc-recommended

See if you can use it

DanilYachmenev commented:
Before I do a review, pls, do the following:
1) Read the [Documentation about guidelines ](https://github.com/sorrentum/sorrentum/blob/master/docs/Documentation_about_guidelines.md)top-to-bottom once again and refactor the doc according to it
    - I see that there are still some crucial actions undone, like running linter
    - Take a look at how this document is formatted itself, look at its code - this one should be done in a very similar manner
2) 1) Read the [First review process](https://github.com/sorrentum/sorrentum/blob/master/docs/First_review_process.md) doc top-to-bottom once again and edit your PR according to it
    - This now refers not to the code, but to the GitHub PR and communication with reviewers
    - This is not less crucial then the code itself, trust me, especially for you
3) Read other docs that we published
    - e.g. the branch is named incorrectly. pls fins a doc information about it is located

Dylan-Celli commented:
I got the following message from running the setenv_amp.sh script, does the no file or directory error mean anything if it is still printing success?

./setenv_amp.sh: line 83: /home/dcelli/sorrentum/dev_scripts/dev_scripts/setenv_amp.configure_env.sh: No such file or directory
==> SUCCESS <==
-------------------------------------------------------------------------------------------------------------------------------------------------
I got the following error when I tried to run the invoke lint command, is this something wrong with my setup?
One and only one set-up config should be true:
is_cmamp_prod=False
is_dev4=False
is_dev_ck=False
is_ig_prod=False
is_inside_ci=False
is_mac=False
Traceback (most recent call last):
  File "/usr/bin/invoke", line 11, in <module>
    load_entry_point('invoke==1.4.1', 'console_scripts', 'invoke')()
  File "/usr/lib/python3/dist-packages/invoke/program.py", line 373, in run
    self.parse_collection()
  File "/usr/lib/python3/dist-packages/invoke/program.py", line 465, in parse_collection
    self.load_collection()
  File "/usr/lib/python3/dist-packages/invoke/program.py", line 696, in load_collection
    module, parent = loader.load(coll_name)
  File "/usr/lib/python3/dist-packages/invoke/loader.py", line 76, in load
    module = imp.load_module(name, fd, path, desc)
  File "/usr/lib/python3.10/imp.py", line 235, in load_module
    return load_source(name, filename, file)
  File "/usr/lib/python3.10/imp.py", line 172, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 719, in _load
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/dcelli/sorrentum/tasks.py", line 131, in <module>
    ECR_BASE_PATH = os.environ["CK_ECR_BASE_PATH"]
  File "/usr/lib/python3.10/os.py", line 679, in __getitem__
    raise KeyError(key) from None
KeyError: 'CK_ECR_BASE_PATH'

DanilYachmenev commented:
> I got the following message from running the setenv_amp.sh script, does the no file or directory error mean anything if it is still printing success?

1) A general code style rule - if you post a code snippet, put it in tripple backticks before and after (```)
This way it will be formatted as smth like
```
line1
line2
...
```
Pls edit your comment following that

2) When posting an error, **always** put a command that you've used so we know what you've called
because quite often the problem is in the request itself

3) I have no `sorrentum/dev_scripts/dev_scripts/setenv_amp.configure_env.sh` either
at least 1 `/dev_scripts` seems extra
try to call the script with a correct path
and post here if any troubles appear, following aforementioned

Dylan-Celli commented:
Ok I fixed the previous edit to include the code block and fixed one of my issues.

This is the command I ran from my sorrentum directory
```
invoke linter --files "./docs/python_unit_tests.md"
```
This was the error that resulted from the command
```

samarth9008 commented:
Were you inside the python environment? The terminal should look like this
`(amp.client_venv) samarthk@dev1:~/src/cmamp1$ `


The correct command for python files is 
`> i lint --files "defi/tulip/implementation/order.py defi/tulip/implementation/order_matching.py"`.

However, for you since it is a markdown file refer to [this](https://github.com/sorrentum/sorrentum/blob/master/dev_scripts/lint_md.sh) file for linting.

Dylan-Celli commented:
> Were you inside the python environment? The terminal should look like this
> `(amp.client_venv) samarthk@dev1:~/src/cmamp1$ `

I was not in the python environment, I was just in the linux CMD


> The correct command for python files is
> `> i lint --files "defi/tulip/implementation/order.py defi/tulip/implementation/order_matching.py"`

I was not able to get the invoke command to work using just i, I always had to spell out invoke

> However, for you since it is a markdown file refer to [this](https://github.com/sorrentum/sorrentum/blob/master/dev_scripts/lint_md.sh) file for linting.

After I ran that script the linter ran, does that suffice for the task of running the linter on my md file?

samarth9008 commented:
> I was not in the python environment, I was just in the linux CMD

Refer to the second reference of my previous comment. There is a reason why we tell you to go through the #189 issue, so that you do not face any kinds of error which you are facing currently. 
 
> I was not able to get the invoke command to work using just i, I always had to spell out invoke

You will get it once you are inside the environment.
 
Refer to the files I have pointed you to.

Dylan-Celli commented:
> @Dylan-Celli we have carefully worked on a way to get people up to speed, so you need to follow the steps carefully.
> 
> In your specific case, your env is not configured properly. `source dev_scripts/setenv_amp.sh` is setting up the aliases including `i` for `invoke`

I did run that script with no errors but 'i' is still not an alias for invoke 
output from when I ran the script
```
# Activate virtual env '/home/dcelli/src/venv/amp.client_venv/bin/activate'
which python=/home/dcelli/src/venv/amp.client_venv/bin/python
python -v=Python 3.10.6
# Set path
PATH=
/home/dcelli/sorrentum/dev_scripts/testing
/home/dcelli/sorrentum/dev_scripts/notebooks
/home/dcelli/sorrentum/dev_scripts/install
/home/dcelli/sorrentum/dev_scripts/infra
/home/dcelli/sorrentum/dev_scripts/git
/home/dcelli/sorrentum/dev_scripts/aws
/home/dcelli/sorrentum/dev_scripts
/home/dcelli/sorrentum
.
/home/dcelli/src/venv/amp.client_venv/bin
/usr/local/sbin
/usr/local/bin
/usr/sbin
/usr/bin
/sbin
/bin
/usr/games
/usr/local/games
/snap/bin
/snap/bin
# Set PYTHONPATH
PYTHONPATH=
/home/dcelli/sorrentum

# Configure env
which gh=
AM_AWS_PROFILE=am
AM_AWS_S3_BUCKET=alphamatic-data
AM_ECR_BASE_PATH=665840871993.dkr.ecr.us-east-1.amazonaws.com
CK_AWS_PROFILE=ck
CK_AWS_S3_BUCKET=cryptokaizen-data
alias i='invoke'
alias ih='invoke --help'
alias il='invoke --list'
alias it='invoke traceback'
alias itpb='pbpaste | traceback_to_cfile.py -i - -o cfile'
==> SUCCESS <==
```

DanilYachmenev commented:
> alias i='invoke'

from your log

samarth9008 commented:
Please read this comment in the edit mode. Edit mode can be activated from the top right 3 dots button.

In the gdoc wherever you see any content between three backticks **```** like this
```
# Run only fast tests:
> i run_fast_tests
# Run only slow tests:
> i run_slow_tests
# Run only superslow tests:
> i run_superslow_tests
```
The whole section should appear inside a grey box (as you saw these comment without the edit mode). We use **>** as a reference for people who is reading this doc to tell them that these commands should be executed inside the terminal.

Let me know if it is unclear. Every starting ``` in the gdoc has an ending ```. Look carefully from the gdoc.

samarth9008 commented:
The invoke section for me would look something like this

## Using invoke
[`invoke`](https://www.pyinvoke.org/) is a task execution which allows to execute some typical workflows, e.g. run the tests
```
# Run only fast tests:
> i run_fast_tests
# Run only slow tests:
> i run_slow_tests
# Run only superslow tests:
> i run_superslow_tests
```
To see the options use `--help` option, e.g. `i --help run_fast_tests`:
```
Usage: inv[oke] [--core-opts] run_fast_tests [--options] [other tasks here ...]

Docstring:
  Run fast tests.

  :param stage: select a specific stage for the Docker image
  :param pytest_opts: option for pytest
  :param skip_submodules: ignore all the dir inside a submodule
  :param coverage: enable coverage computation
  :param collect_only: do not run tests but show what will be executed
  :param tee_to_file: save output of pytest in `tmp.pytest.log`
  :param kwargs: kwargs for `ctx.run`

Options:
  -c, --coverage
  -k, --skip-submodules
  -o, --collect-only
  -p STRING, --pytest-opts=STRING
  -s STRING, --stage=STRING
  -t, --tee-to-file
  -v STRING, --version=STRING
```

DanilYachmenev commented on docs/python_unit_tests_figs/image_1.png:
Images should not be commited
read https://github.com/sorrentum/sorrentum/blob/master/docs/Documentation_about_guidelines.md#embedding-screenshots

gpsaggese replied:
@DanilYachmenev sorry we have changed approach since the new flow https://github.com/sorrentum/sorrentum/blob/master/docs/Documentation_about_guidelines.md#gdocs---markdown-using-pandoc-recommended automatically creates files ... 

Can you pls update the documentation accordingly?

Dylan-Celli replied:
So, when I added the images to the folder I did as GP about it first and [he told me](https://github.com/sorrentum/sorrentum/issues/300#issuecomment-1574444106) to add the images in a local folder. Does that suffice or is there something else that I should be doing.

Dylan-Celli commented:
> In the gdoc wherever you see any content between three backticks **``**

I went through and cleaned up the md to look like the doc more. Is there anything else that I am missing?

samarth9008 commented on docs/python_unit_tests.md:
```diff
@@ -1,3 +1,64 @@
+- [Running tests on GH Actions](#running-tests-on-gh-actions)
+  * [How to run a single test on GH Action](#how-to-run-a-single-test-on-gh-action)
+- [Guidelines about writing unit tests](#guidelines-about-writing-unit-tests)
+    + [**What is a unit test?**](#--what-is-a-unit-test---)
```
Use the tool lint_md.sh to generate Table of Contents automatically.

Dylan-Celli replied:
I did run the lint_md on the file but it did not generate the table

DanilYachmenev commented on docs/python_unit_tests.md:
```diff
@@ -0,0 +1,1408 @@
+      things are not horrible broken
+  - `slow`
+    - Tests that we don't want to run all the times because they are:
+      - Slow (typically &lt; 20 seconds per test)
```
nah, slow tests are 6-20 seconds
> 20 are superslow

Dylan-Celli commented on docs/python_unit_tests.md:
```diff
@@ -0,0 +1,1408 @@
+   - Rationale: we want each test to be clear, simple, fast
+   - If there is repeated code we should factor it out (e.g., builders for
+     objects)
+6) Each test should be crystal clear on how is different from the others
```
I fixed this in the most recent commit.

DanilYachmenev commented:
> @DanilYachmenev Can you please review it?

sure, will do today, thx for the ping

@Dylan-Celli read https://github.com/sorrentum/sorrentum/blob/master/docs/First_review_process.md as suggested before and treat your PR accordingly
for example, this PR had no label, so when I was going through the lib, I did not consider that your PR was ready for review

request notifications can be omitted, labels are not

DanilYachmenev commented on dev_scripts/setenv_amp.sh:
@samarth9008 I believe you've removed this file
is there a reason for that?

samarth9008 replied:
It was an empty file.

Dylan-Celli replied:
The reason why Samarth removed the setenv_amp.sh file was because I had accidentally made an empty commit on that file and he did not want it to be reflected on the main branch when merging.

DanilYachmenev replied:
@samarth9008 @Dylan-Celli so guys I believe we should bring this file back
because currently it is going to be simply removed from the master branch
and we should not let it happen since his file is crucial for setting up the env
 
<img width="1418" alt="Снимок экрана 2023-06-22 в 17 31 19" src="https://github.com/sorrentum/sorrentum/assets/31514660/005d0a64-2190-4f7e-acb9-8c15307263e1">

samarth9008 replied:
@DanilYachmenev Understand your point but wouldn't it also be a problem if we commit an empty file? I think it will be replaced in the master branch.

@Dylan-Celli Make sure you only commit those files which are required for this PR.

DanilYachmenev replied:
well, should we commit an empty file at all?
Let me know if it is somehow necessary for the task
but for now it seems that in master this file is not empty so we can revert it back as it is
the goal is to get rid of any commited changes to this file

@Dylan-Celli if it is hard to revert, try to recreate it on your branch with same name and content
and make sure it disappears from the changed files in this PR after the push

Dylan-Celli replied:
Ok so just make a new branch with the same name to see if it overwrite the current branch and commit the md file from there?

DanilYachmenev reviewed (commented):
Pretty good! see some refinement nits but overall really close to merge

pls go over some general suggestions that you can check easily (like invoking all the fast tests) to check if it works
Ask questions in case smth is unclear

samarth9008 commented:
@Dylan-Celli Can you implement the changes so we can move one step further to merge?

Dylan-Celli commented:
> move

Done! 👍

Dylan-Celli commented:
I added the suggestions, I did not see any `pytest.raises` in the markdown

DanilYachmenev commented on docs/unit_tests.md:
````diff
@@ -0,0 +1,1408 @@
+might be useful when a user wants to run regressions on the local Docker image
+to verify that nothing is broken before promoting it to `dev` image.
+
+```
+> i run_fast_tests --stage local
+```
````
indent all the code blocks in the file correctly
https://github.com/sorrentum/sorrentum/blob/master/docs/Documentation_about_guidelines.md#indenting-code-style

DanilYachmenev commented on docs/unit_tests.md:
```diff
@@ -0,0 +1,1408 @@
+# Unit tests
+- [Running unit tests](#running-unit-tests)
```
TOC should be created using `dev_scripts/lint_md.sh`
https://github.com/sorrentum/sorrentum/blob/master/docs/Documentation_about_guidelines.md#table-of-content-toc

Dylan-Celli replied:
I found an issue with the `linter_md.sh` script
when adding the following headers into the TOC
```
Tip #1: Test one thing
Tip #2: Keep tests self-contained
Tip #3: Only specify data related to what is being tested
Tip #4: Test realistic corner cases
Tip #5: Test a typical scenario
Tip #6: Test executable scripts end-to-end
```
it generated the following links
```
+ [Tip #1: Test one thing](#tip-%231-test-one-thing)
    + [Tip #2: Keep tests self-contained](#tip-%232-keep-tests-self-contained)
    + [Tip #3: Only specify data related to what is being tested](#tip-%233-only-specify-data-related-to-what-is-being-tested)
    + [Tip #4: Test realistic corner cases](#tip-%234-test-realistic-corner-cases)
    + [Tip #5: Test a typical scenario](#tip-%235-test-a-typical-scenario)
    + [Tip #6: Test executable scripts end-to-end](#tip-%236-test-executable-scripts-end-to-end)
```
Whenever it finds a `#` after the initial `#` that denotes heading level it replaces it with `23`

EX:
Header: `Tip #2: Keep tests self-contained`
Generated link: `#tip-%232-keep-tests-self-contained`
Actual link: `#tip-2-keep-tests-self-contained`

DanilYachmenev replied:
well, in this case lets just remove `Tip #1: `-like additions
thx for the note, but lets not over-optimize here

Dylan-Celli replied:
I corrected the links so they work as is. Do you still want me to reformat the headers or can I leave them if the links work

DanilYachmenev replied:
yes, in this case `Tip` word is extra

DanilYachmenev commented on docs/unit_tests.md:
````diff
@@ -0,0 +1,1408 @@
+
+```
+
+This generates and run a pytest command inside Docker like:
````
code commands like `pytest` should be put in backticks

DanilYachmenev commented on docs/unit_tests.md:
````diff
@@ -0,0 +1,1408 @@
+docker> /venv/bin/pytest -m "not slow and not superslow" oms/test/test_broker.py::TestSimulatedBroker1 --cov=. --cov-branch --cov-report term-missing --cov-report html
+```
+
+- which generates:
````
Start all the bullet points with capital letters for consistency
do for the whole file

DanilYachmenev commented on docs/unit_tests.md:
start file name from capital U

DanilYachmenev commented on docs/unit_tests.md:
```diff
@@ -0,0 +1,1408 @@
+  - Avoid many files with a lot of boilerplate code
+- Cons of 1) vs 2)
+  - The single file can become huge!
+- Compromise solution: _ Start with a single file `test_$DIRNAME.py` (or
```
remove trailing `_`, they do not look good

DanilYachmenev commented on docs/unit_tests.md:
```diff
@@ -0,0 +1,1408 @@
+  - The single file can become huge!
+- Compromise solution: _ Start with a single file `test_$DIRNAME.py` (or
+  `test_dir_name.py`) _ In the large file add a framed comment like: \
+  '# ##################'\
+  '# Unit tests for file: … '\
+  '# ##################' _ 
```
put these in tripple backticks as a code block instead of putting each line in single backticks
```
# ##################
# Unit tests for file: … 
# ##################
```

DanilYachmenev commented on docs/unit_tests.md:
```diff
@@ -0,0 +1,1408 @@
+- It is OK to use non-private methods in test classes to ensure that the code is
+  in order of dependency, so that the reader doesn't have to jump back / forth
+- We want to separate chunks of unit test code using:
+  ` # //////////////////////////////////////////////////////////////////////// `
```
replace all the slashes with `#` here so
`# #######...`

DanilYachmenev commented on docs/unit_tests.md:
````diff
@@ -0,0 +1,1408 @@
+      def test...2(self):
+          ...
+  ```
+- Each TestCase tests a "behavior" like a set of related methods
````
`TestCase` should be backticked

DanilYachmenev commented on docs/unit_tests.md:
```diff
@@ -0,0 +1,1408 @@
+
+## Some general suggestions about testing
+
+1) Test from the outside-in
```
replace number with tripple ### in order to make them bullet pints, reflected in TOC

Dylan-Celli replied:
I implemented the change but `lint_md.sh` is not adding it the TOC. Should I just add them manually?

DanilYachmenev reviewed (commented):
See another round of comments
pls check the preview of your file after you make changes and make sure it looks good visually
use existing documentation as reference

and another reason to read https://github.com/sorrentum/sorrentum/blob/master/docs/Documentation_about_guidelines.md again

attention to details is a crucial thing that we are accessing and you need to demonstrate it here

samarth9008 commented:
Shall we close this PR @Dylan-Celli ?

DanilYachmenev commented:
@Dylan-Celli finished up the doc myself due to the high priority
see the implemented changes

DanilYachmenev commented on dev_scripts/setenv_amp.sh:
```diff
@@ -82,4 +82,4 @@ echo $PYTHONPATH | perl -e 'print join("\n", grep { not $seen{$_}++ } split(/:/,
 source $AMP/dev_scripts/setenv_amp.configure_env.sh
 
-echo "==> SUCCESS <=="
+echo "==> SUCCESS <=="
```
not breaking anything, recreated the file, env is being set correctly

DanilYachmenev commented:
@gpsaggese merging to unblock proceeding with other tasks
will implement any needed comments in a follow-up
